x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/GrokPatternCreator.java

```diff
@@ -0,0 +1,600 @@
+            if (fieldName == null) {
+                throw new IllegalStateException("Cannot process KV matches until a field name has been determined");
+            }
+            Grok grok = new Grok(Grok.getBuiltinPatterns(), "(?m)%{DATA:" + PREFACE + "}\\b" +
+                fieldName + "=%{USER:" + VALUE + "}%{GREEDYDATA:" + EPILOGUE + "}");
+            Collection<String> values = new ArrayList<>();
+            for (String snippet : snippets) {
+                Map<String, Object> captures = grok.captures(snippet);
+                // If the pattern doesn't match then captures will be null
+                if (captures == null) {
+                    throw new IllegalStateException("[\\b" + fieldName + "=%{USER}] does not match snippet [" + snippet + "]");
+                }
+                prefaces.add(captures.getOrDefault(PREFACE, "").toString());
+                values.add(captures.getOrDefault(VALUE, "").toString());
+                epilogues.add(captures.getOrDefault(EPILOGUE, "").toString());
+            }
+            String adjustedFieldName = buildFieldName(fieldNameCountStore, fieldName);
+            if (mappings != null) {
+                mappings.put(adjustedFieldName, AbstractLogFileStructureFinder.guessScalarMapping(explanation, adjustedFieldName, values));
+            }
+            return "\\b" + fieldName + "=%{USER:" + adjustedFieldName + "}";
+        }
+    }
+
+    /**
+     * A Grok pattern candidate that matches a single named Grok pattern but will not update mappings.
+     */
+    static class NoMappingGrokPatternCandidate extends ValueOnlyGrokPatternCandidate {
+
+        NoMappingGrokPatternCandidate(String grokPatternName, String fieldName) {
+            super(grokPatternName, null, fieldName);
+        }
+
+        @Override
+        public String processCaptures(Map<String, Integer> fieldNameCountStore, Collection<String> snippets, Collection<String> prefaces,
+                                      Collection<String> epilogues, Map<String, Object> mappings) {
+            return super.processCaptures(fieldNameCountStore, snippets, prefaces, epilogues, null);
+        }
+    }
+
+    /**
+     * Used to check whether a single Grok pattern matches every sample message in its entirety.
+     */
+    static class FullMatchGrokPatternCandidate {
+
+        private final String grokString;
+        private final String timeField;
+        private final Grok grok;
+
+        FullMatchGrokPatternCandidate(String grokPatternName, String timeField) {
+            grokString = "%{" + grokPatternName + "}";
+            this.timeField = timeField;
+            grok = new Grok(Grok.getBuiltinPatterns(), grokString);
+        }
+
+        public boolean matchesAll(Collection<String> sampleMessages) {
+            return sampleMessages.stream().allMatch(grok::match);
+        }
+
+        /**
+         * This must only be called if {@link #matchesAll} returns <code>true</code>.
```

Comment:
As this is only called in one place and right after `matchesAll` was called, I think we could move the `matchesAll` check in this method and so remove the risk of letting it be called without fulfilling the `matchesAll` contract.